### PR TITLE
[WIP] 1.8 Profiler comments

### PR DESCRIPTION
**Pull request by carmocca**
Please ignore 🙈 

This is for CI and addressing comments for #6067

cc: @tchaton to address these and maybe merge

### CODE REVIEW

File: pytorch_lightning/profiler/pytorch.py [hunk 1]
```diff
@@ -159,18 +104,36 @@ def __init__(
...
+            ValueError:
+                If you attempt to stop recording an action which was never started.
         """
+        if profiled_functions is not None:
+            # TODO: DeprecationWarning
+            if record_functions is None:
+                record_functions = profiled_functions
+            else:
+                # TODO; raise MisconfigurationException
+                pass
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:135]: `profiled_functions` was renamed to `record_functions`. Should we do this?

File: pytorch_lightning/profiler/pytorch.py [hunk 2]
```diff
@@ -237,7 +200,8 @@ def on_train_start(self, local_rank: Optional[str] = None, log_dir: str = None):
         streaming_out = [self.output_file.write] if self.output_file else [log.info]
         super().__init__(output_streams=streaming_out)
 
-    def wrap_functions_into_rank_zero_only(self):
+    def wrap_functions_into_rank_zero_only(self) -> None:
+        # TODO: what's the point of wrapping like this?
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:204]: Address this

File: pytorch_lightning/profiler/pytorch.py [hunk 3]
```diff
@@ -253,16 +217,16 @@ def start(self, action_name: str) -> None:
...
-            self._create_profiler(action_name, torch.cuda.profiler.profile, enter=False)
-            self._create_profiler(action_name, torch.autograd.profiler.emit_nvtx)
+            self._create_profiler(torch.cuda.profiler.profile, enter=False)
+            self._create_profiler(torch.autograd.profiler.emit_nvtx)
         else:
-            self._create_profiler(action_name, torch.autograd.profiler.profile)
+            self._create_profiler(torch.autograd.profiler.profile)
 
-    def _create_profiler(self, action_name, profiler, enter=True):
+    def _create_profiler(self, profiler, enter=True):
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:229]: `action_name` was unused so removed it. Maybe it should be used?

File: pytorch_lightning/profiler/pytorch.py [hunk 4]
```diff
@@ -323,101 +281,55 @@ def summary(self) -> str:
...
+                path_to_trace = (
+                    filename
+                    if self.path_to_export_trace is None else
+                    os.path.join(self.path_to_export_trace, filename)
+                )
                 function_events.export_chrome_trace(path_to_trace)
 
             if self.emit_nvtx:
-                return output_string
+                return recorded_stats
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:292]: Shouldn't this be outside of the `profiled_actions` loop?

File: pytorch_lightning/profiler/pytorch.py [hunk 5]
```diff
@@ -468,23 +380,129 @@ class PyTorchProfiler(LegacyPyTorchProfiler):
...
+        def __enter__(self):
+            for module_name, module in self._model.named_modules():
+                key = module_name if module in self.NN_MODULES else f"{type(module)}: {module_name}"
+                pre_handle = module.register_forward_pre_hook(partial(self._start_recording, key=key))
+                post_handle = module.register_forward_hook(partial(self._stop_recording, key=key))
+                self._handles[module_name] = [pre_handle, post_handle]
+
+        def __exit__(self, *_):
+            self._records = {}
+            self._handles = {}
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:447]: Is this not enough to free up the memory?

File: pytorch_lightning/profiler/pytorch.py [hunk 6]
```diff
@@ -531,41 +551,41 @@ def __init__(
...
-            self.sort_by_key = sort_by_key
+                on_trace_ready: Function which takes the profiler and executed on ``RECORD_AND_SAVE`` action
+
+                export_to_flame_graph: Whether to export the sequence of profiled operators for Flame Graph.
+
+                with_flops: TODO
+
+                metric: TODO
 
+                record_module_names: TODO
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:587]: See these

File: pytorch_lightning/profiler/pytorch.py [hunk 7]
```diff
@@ -607,40 +628,26 @@ def __init__(
             self.context_names = {}
             self.running_stack = []
             self.profiler = None
-            self.lightning_module = None
+            self.lightning_module = None  # TODO: where is this set?
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:631]: See TODO

File: pytorch_lightning/profiler/pytorch.py [hunk 8]
```diff
@@ -207,7 +170,7 @@ def __init__(
             self.on_train_start(local_rank=local_rank)
             self.on_train_start = super().on_train_start
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:171]: I don't understand the purpose of this. Why do we define `on_train_start` in this class if we are going to set it to the parents? Maybe the `on_train_start` function in this class should have a different name

File: pytorch_lightning/profiler/pytorch.py [hunk 9]
```diff
@@ -607,40 +628,26 @@ def __init__(
...
             self.profiler = None
-            self.lightning_module = None
+            self.lightning_module = None  # TODO: where is this set?
             self.register = None
 
             self.output_fname = output_filename
             self.output_file = None
             if local_rank is not None:
-                super().on_train_start(local_rank=local_rank)
+                super().on_train_start(local_rank=local_rank)  # TODO: why super()?
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:637]: Shouldn't this call `self.on_train_start`?

File: pytorch_lightning/profiler/pytorch.py [hunk 10]
```diff
@@ -607,40 +628,26 @@ def __init__(
...
-
+        def _summary(self, local_rank: int) -> dict:
             self.profiler.__exit__(None, None, None)
             if self.register is not None:
                 self.register.__exit__(None, None, None)
 
             data = self.profiler.events().key_averages(group_by_input_shapes=self.group_by_input_shapes)
             table = data.table(sort_by=self.sort_by_key, row_limit=self.row_limit)
-            recorded_stats[self.START_ACTION] = table
+            return {self.START_ACTION: table}
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:647]: Does it make sense that this returns just one action?

File: pytorch_lightning/profiler/pytorch.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:135]: It was to uniformize the API between both 1.8 and before.

File: pytorch_lightning/profiler/pytorch.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:204]: The function should be run only on rank zero if `output_fname` is not profiled. 
If running with DDP and the user provides `output_fname`, each process will save their own profiling into `output_fname_{rank}.txt`

File: pytorch_lightning/profiler/pytorch.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:229]: Fine :)

File: pytorch_lightning/profiler/pytorch.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:292]: Yes, could be moved with self.enabled.

File: pytorch_lightning/profiler/pytorch.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:447]: good question. I am not sure the handles remove the hook if deleted.

File: pytorch_lightning/profiler/pytorch.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:587]: Yes.

File: pytorch_lightning/profiler/pytorch.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:631]: It is set by the ProfilerConnector to enable recording on function names.

File: pytorch_lightning/profiler/pytorch.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:171]: The user could want to use the profiler without Lightning and the functions need to be executed if the user provides a rank.

File: pytorch_lightning/profiler/pytorch.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:637]: Yes, should be the same here. I overrided it and forgot to remove this part.

File: pytorch_lightning/profiler/pytorch.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:647]: It is being logged in the terminal. This action covers the entire training / validation.

File: pytorch_lightning/profiler/pytorch.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:135]: I get that but we should deprecate then

File: pytorch_lightning/profiler/pytorch.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:204]:
```suggestion
```

File: pytorch_lightning/profiler/pytorch.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:631]:
```suggestion
            self.lightning_module = None  # set by ProfilerConnector
```

File: pytorch_lightning/profiler/pytorch.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:171]: I see. In any case, we could rename it to `initialize` so there's no need to hack the hooks

File: pytorch_lightning/profiler/pytorch.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:647]: So if this supported other calls like `.test` then would it have more actions?

File: pytorch_lightning/profiler/pytorch.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/profiler/pytorch.py:446]: Lets be safe

```suggestion
            for h in self._handles.values():	
                h.remove()  # remove module forward hooks
            self._records = {}
            self._handles = {}
```

File: pytorch_lightning/profiler/pytorch.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/profiler/pytorch.py:648]: Currently doesn't but yes it should.

**Pull request merged**